compensations/src/main/java/org/jboss/narayana/compensations/impl/TransactionImpl.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.narayana.compensations.impl;
+
+import org.jboss.narayana.compensations.api.Action;
+import org.jboss.narayana.compensations.api.Compensatable;
+import org.jboss.narayana.compensations.api.CompensationHandler;
+import org.jboss.narayana.compensations.api.ConfirmationHandler;
+import org.jboss.narayana.compensations.api.Transaction;
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+public class TransactionImpl implements Transaction {
+
+    private List<ActionInfo> actions = new LinkedList<>();
+
+    @Override
+    public Transaction doAction(Action action) {
+        return doAction(action, null, null);
+    }
+
+    @Override
+    public Transaction doAction(Action action,
+            CompensationHandler compensationHandler) {
+
+        return doAction(action, compensationHandler, null);
+    }
+
+    @Override
+    public Transaction doAction(Action action,
+            ConfirmationHandler confirmationHandler) {
+
+        return doAction(action, null, confirmationHandler);
+    }
+
+    @Override
+    public Transaction doAction(Action action,
+            CompensationHandler compensationHandler,
+            ConfirmationHandler confirmationHandler) {
+
+        actions.add(new ActionInfo(action, compensationHandler, confirmationHandler));
+
+        return this;
+    }
+
+    @Override
+    @Compensatable
+    public void execute() {
+        actions.forEach(actionInfo -> {
+            Set<ParticipantManager> managers = enlist(
+                    actionInfo.compensationHandler,
+                    actionInfo.confirmationHandler);
+            try {
+                actionInfo.action.execute();
+            } catch (RuntimeException e) {
+                exit(managers);
```

Comment:
Since Compensations API is based on JTA 1.2, they have the same exception handling semantics. Here's the extract from the spec:

> By default checked exceptions do not result in the transactional interceptor
> marking the transaction for rollback and instances of RuntimeException and its
> subclasses do.